src/text-buffer.coffee

```diff
@@ -1138,7 +1172,22 @@ class TextBuffer
       else
         iterator = new MatchIterator.ForwardsMultiLine(this, regex, range)
 
-    iterator.iterate(callback, global)
+    callback_with_context = (argument) =>
+      argument.linesBefore = []
+      for i in [0...(options.lineCountBefore or 0)]
+        row = argument.range.start.row - i - 1
+        break if row < 0
+        argument.linesBefore.unshift(@lineForRow(row))
```

Comment:
Since `unshift` has to move every element in the array, I think it'd be slightly more direct and fast to iterate through the buffer's lines in the *forwards* direction. Maybe something like:

```coffee
row = Math.max(0, range.start.row - lineCountBefore)
while row < range.start.row
  argument.linesBefore.push(@lineForRow(row))
```